text/0000-serialization-format.md

```diff
@@ -0,0 +1,125 @@
+users, as there is no way of falling back to load previously serialized states.
+
+This RFC is going to address this issue by introducing a `SerializationHeader`
+message and an interface for parsing serialized states, independent of the
+serialization format.
+
+## Approach
+
+[approach]: #approach
+
+For being able to identify the underlying serialization format, we introduce a
+`SerializationHeader` message.
+
+```
+message SerializationHeader {
+  uint32 format_version = 1;
+}
+```
+
+We further change the current `SerializedData` protobuf message, which is meant
+to be able to contain `oneof` any of the message types required for serializing
+the mapping state:
+
+```
+message SerializedData {
+  oneof data {
+    PoseGraph pose_graph = 1;
+    AllTrajectoryBuilderOptions all_trajectory_builder_options = 2;
+    Submap submap = 3;
+    TrajectoryData trajectory_data = 4;
+    Node node = 5;
+    ImuData imu_data = 6;
+    OdometryData odometry_data =7;
+    FixedFramePoseData fixed_frame_pose_data = 8;
+    LandmarkData landmark_data = 9;
+  }
+}
+```
+
+~~For backwards compatibility, we will rename the original `SerializedData`
+definition to `LegacySerializedData`.~~ (see discussion)
+
+By keeping all the payload related data wrapped in a single message type, we can
+also easily switch to a different underlying storage format in the future (e.g.
+[Riegeli](https://github.com/google/riegeli/)).
+
+### Generic Serialization Order
+
+Using the above messages, the generic serialization order is defined as:
+
+*Generic Serialization Order* |
+:---------------------------: |
+SerializationHeader           |
+(SerializedData)*             |
+
+### Version 1.0 Serialization Order
+
+Using the definitions above, the proposed serialization format for version 1.0
+is as follows
+
+*Version 1.0 Serialization Order* |
+:-------------------------------: |
+SerializationHeader               |
+PoseGraph                         |
+AllTrajectoryBuilderOptions       |
+(Submap)\*                        |
+(Node)\*                          |
+(ImuData)\*                       |
+(OdometryData)\*                  |
+(FixedFramePoseData)\*            |
+(TrajectoryData)\*                |
+(LandmarkData)\*                  |
+
+### Backwards-Compatibility
+
+*   ~~ To gracefully transition to the new format, we will provide an internal
+    fallback to the *old* parsing of PbStreams. Meaning, we will first try to
+    read the a `SerializationHeader` from the pbstream and if this fails, try
+    parsing the *old* format. This way we can at least provide functionality for
+    parsing the last pre-header version. ~~ (see discussion)
+*   We will provide a migration tool to re-write already serialized pbstreams
+    into the new format.
+
+## Discussion Points
+
+[discussion]: #discussion
+
+*   Version string for new format: 0.9 or 1.0?
+    -   `uint32` - first version == 1
+*   Duplicating protos to separate storage and *wire* format following
```

Comment:
I do not understand this sentence. What is it trying to say?

Reply:
I only wanted to mention this as I have read about this being a best practice.
It seems the advice is to always have two pairs of protos: One for what you use for storage and one which you use e.g. in RPC calls or also within your running system in objects. This allows you to change your storage format without the rest of your code. Right now it looks like we use the same for both.

I also already discussed with @cschuet that we should probably postpone and discuss this separately if it comes up again.